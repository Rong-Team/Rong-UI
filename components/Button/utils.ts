import { ButtonGroupConfig } from "./btn-group-context"
import { BaseButtonProps, NativeButtonProps } from "./button"

export const filterPropsWithGroup = (
    props: React.PropsWithChildren<NativeButtonProps>,
    config: ButtonGroupConfig,
  ): NativeButtonProps => {
    if (!config.isButtonGroup) return props
    return {
      ...props,
      shape:undefined,
      block:false,
      size: config.size || props.size,
    }
  }


  