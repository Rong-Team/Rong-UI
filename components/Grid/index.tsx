import React from 'react'
import GridContainer from './container'
import GridItem,{GridBasicItemProps} from './items'
interface Props extends React.FC<GridBasicItemProps> {
    Container:typeof GridContainer
}

const Grid:Props=({children,...props})=>{
    return <GridItem {...props}>
            {children}
    </GridItem>
}
Grid.Container=GridContainer

export default Grid
