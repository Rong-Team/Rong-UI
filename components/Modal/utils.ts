function getScroll(w: Window, top?: boolean): number {
    let ret :any = w[`page${top ? 'Y' : 'X'}Offset` as any];
    const method:any = `scroll${top ? 'Top' : 'Left'}`;
    if (typeof ret !== 'number') {
      const d = w.document;
      ret = d.documentElement[top?'scrollTop':'scrollLeft'];
      if (typeof ret !== 'number') {
        ret = d.body[top?'scrollTop':'scrollLeft']  ;
      }
    }
    return ret;
  }
  
  interface CompatibleDocument extends Document {
    parentWindow?: Window;
  }
  
  export function offset(el: Element) {
    const rect = el.getBoundingClientRect();
    const pos = {
      left: rect.left,
      top: rect.top,
    };
    const doc = el.ownerDocument as CompatibleDocument;
    const w = doc.defaultView || doc.parentWindow;
    pos.left += getScroll(w as any);
    pos.top += getScroll(w as any, true);
    return pos;
  }