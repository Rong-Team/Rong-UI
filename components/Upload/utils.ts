import { UploadFile, RFile } from "./interface";

export function fileToObject(file: RFile): UploadFile {
    return {
        ...file,
        lastModified: file.lastModified,
        lastModifiedDate: file.lastModifiedDate,
        name: file.name,
        size: file.size,
        type: file.type,
        uid: file.uid,
        percent: 0,
        originFileObj: file,
    } as UploadFile;
}

export function getFileItem(file: UploadFile, fileList: UploadFile[]) {
    const matchKey = file.uid !== undefined ? 'uid' : 'name';
    return fileList.filter(item => item[matchKey] === file[matchKey])[0];
}

export function removeFileItem(file: UploadFile, fileList: UploadFile[]) {
    const matchKey = file.uid !== undefined ? 'uid' : 'name';
    const removed = fileList.filter(item => item[matchKey] !== file[matchKey]);
    if (removed.length === fileList.length) {
        return null;
    }
    return removed;
}
const extname = (url: string = '') => {
    const temp = url.split('/');
    const filename = temp[temp.length - 1];
    const filenameWithoutSuffix = filename.split(/#|\?/)[0];
    return (/\.[^./\\]*$/.exec(filenameWithoutSuffix) || [''])[0];
};

const isImageFileType = (type: string): boolean => type.indexOf('image/') === 0;

export const isImageUrl = (file: UploadFile): boolean => {
    if (file.type && !file.thumbUrl) {
        return isImageFileType(file.type);
    }
    const url: string = (file.thumbUrl || file.url) as string;
    const extension = extname(url);
    if (
        /^data:image\//.test(url) ||
        /(webp|svg|png|gif|jpg|jpeg|jfif|bmp|dpg|ico)$/i.test(extension)
    ) {
        return true;
    }
    if (/^data:/.test(url)) {
        // other file types of base64
        return false;
    }
    if (extension) {
        // other file types which have extension
        return false;
    }
    return true;
};


const now = +new Date();
let index = 0;

export  function getUid() {
  // eslint-disable-next-line no-plusplus
  return `rong-upload-${now}-${++index}`;
}



interface InternalDataTransferItem extends DataTransferItem {
  isFile: boolean;
  file: (cd: (file: RFile & { webkitRelativePath?: string }) => void) => void;
  createReader: () => any;
  fullPath: string;
  isDirectory: boolean;
  name: string;
  path: string;
}

function loopFiles(item: InternalDataTransferItem, callback:any) {
  const dirReader = item.createReader();
  let fileList:any[] = [];

  function sequence() {
    dirReader.readEntries((entries: Array<InternalDataTransferItem>) => {
      const entryList = Array.prototype.slice.apply(entries);
      fileList = fileList.concat(entryList);

      // Check if all the file has been viewed
      const isFinished = !entryList.length;

      if (isFinished) {
        callback(fileList);
      } else {
        sequence();
      }
    });
  }

  sequence();
}

export const traverseFileTree = (files: Array<InternalDataTransferItem>, callback:any, isAccepted:any) => {
  // eslint-disable-next-line @typescript-eslint/naming-convention
  const _traverseFileTree = (item: InternalDataTransferItem, path?: string) => {
    // eslint-disable-next-line no-param-reassign
    item.path = path || '';
    if (item.isFile) {
      item.file(file => {
        if (isAccepted(file)) {
          if (item.fullPath && !file.webkitRelativePath) {
            Object.defineProperties(file, {
              webkitRelativePath: {
                writable: true,
              },
            });
            // eslint-disable-next-line no-param-reassign
            file.webkitRelativePath = item.fullPath.replace(/^\//, '');
            Object.defineProperties(file, {
              webkitRelativePath: {
                writable: false,
              },
            });
          }
          callback([file]);
        }
      });
    } else if (item.isDirectory) {
      loopFiles(item, (entries: Array<InternalDataTransferItem>) => {
        entries.forEach(entryItem => {
          _traverseFileTree(entryItem, `${path}${item.name}/`);
        });
      });
    }
  };
  files.forEach(file => {
    _traverseFileTree(file.webkitGetAsEntry());
  });
};

function endsWith(str: string, suffix: string) {
    return str.indexOf(suffix, str.length - suffix.length) !== -1;
  }
  
export const attrAccept= (file: RFile, acceptedFiles: string | Array<string>) => {
    if (file && acceptedFiles) {
      const acceptedFilesArray = Array.isArray(acceptedFiles)
        ? acceptedFiles
        : acceptedFiles.split(',');
      const fileName = file.name || '';
      const mimeType = file.type || '';
      const baseMimeType = mimeType.replace(/\/.*$/, '');
  
      return acceptedFilesArray.some(type => {
        const validType = type.trim();
        if (validType.charAt(0) === '.') {
          return endsWith(fileName.toLowerCase(), validType.toLowerCase());
        }
        if (/\/\*$/.test(validType)) {
          // This is something like a image/* mime type
          return baseMimeType === validType.replace(/\/.*$/, '');
        }
        return mimeType === validType;
      });
    }
    return true;
  };

