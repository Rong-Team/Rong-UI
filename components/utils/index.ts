export * from './themeTypes'
export * from './type'
export * from './pick'
export * from './expand'