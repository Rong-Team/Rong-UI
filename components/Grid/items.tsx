import { Justify, Direction, AlignItems, AlignContent } from "./types"
import React from 'react'
import { GridItemWrap } from "./wrapper"
type BreakpointsValue = number | boolean
interface Props {
  xs?: BreakpointsValue
  sm?: BreakpointsValue
  md?: BreakpointsValue
  lg?: BreakpointsValue
  xl?: BreakpointsValue
  justify?: Justify
  direction?: Direction
  alignItems?: AlignItems
  alignContent?: AlignContent
  className?: string
}


type NativeAttrs = Omit<React.HTMLAttributes<any>, keyof Props>

export type GridBasicItemProps = Props  & NativeAttrs
export type ItemLayoutValue = {
  grow: number
  width: string
  basis: string
  display: string
}
const getItemLayout = (val: BreakpointsValue): ItemLayoutValue => {
  const display = val === 0 ? 'display: none;' : ''
  if (typeof val === 'number') {
    const width = (100 / 24) * val
    const ratio = width > 100 ? '100%' : width < 0 ? '0' : `${width}%`
    return {
      grow: 0,
      display,
      width: ratio,
      basis: ratio,
    }
  }
  return {
    grow: 1,
    display,
    width: '100%',
    basis: '0',
  }
}

const GridItem:React.FC<GridBasicItemProps>=(props)=>{
  const {xs=false as BreakpointsValue,
  sm=false as BreakpointsValue,
  md=false as BreakpointsValue,
  lg=false as BreakpointsValue,
  xl=false as BreakpointsValue,
  justify,
  direction,
  alignItems,
  alignContent,
  children,
  className="",
  ...rest
}=props

const getType=():'xs'|'sm'|'md'|'lg'|'xl'=>{

  return 'xs'
}
 
  
  const layout=() => ({
    xs: getItemLayout(xs),
    sm: getItemLayout(sm),
    md: getItemLayout(md),
    lg: getItemLayout(lg),
    xl: getItemLayout(xl),
  })
  return (<GridItemWrap type={getType()} layout={layout()} justify={justify} direction={direction} alignContent={alignContent} alignItems={alignItems}>
    {children}
  </GridItemWrap>)
}

export default GridItem