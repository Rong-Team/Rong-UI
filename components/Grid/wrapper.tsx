import styled, { css } from "styled-components";
import { ItemLayoutValue } from "./items";
import { Wrap } from "./types";

export const GridItemWrap=styled.div<{justify?:string, direction?:string,
    alignItems?:string,
    alignContent?:string,
    layout:{ [key in ['xs', 'sm', 'md', 'lg', 'xl'][number]]: ItemLayoutValue},
    type:'xs'|'sm'|'md'|'lg'|'xl'}>`

    ${props=>props.justify?css`
    justify-content: ${props.justify};
    `:null};
    ${props=>props.direction?css`
   flex-direction: ${props.direction};
    `:null};
    ${props=>props.alignContent?css`
    align-content: ${props.alignContent};
    `:null}
    ${props=>props.alignItems?css`
    align-items: ${props.alignItems};
    `:null}
  
    flex-grow: ${props=>props.layout[props.type].grow};
    max-width: ${props=>props.layout[props.type].width};
    flex-basis: ${props=>props.layout[props.type].basis};

 `

export const GridContainerWrap=styled.div<{wrap:Wrap,unit:string}>`
    display: flex;
    flex-wrap: ${props=>props.wrap};
    box-sizing: border-box;
    margin: calc(-1 * ${props=>props.unit});
    width: calc(100% + ${props=>props.unit} * 2);

`