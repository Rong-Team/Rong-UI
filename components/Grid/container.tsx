import React from 'react'
import {GridBasicItemProps} from './items'
import { Wrap } from './types';
import { GridContainerWrap } from './wrapper';
interface Props {
    gap?: number
    wrap?: Wrap
    className?: string
  }

export type GridContainerProps = Props & GridBasicItemProps

const GridContainer:React.FC<React.PropsWithChildren<GridContainerProps>>=(props)=>{
    const {gap=0,
        wrap='wrap',
        children,
        className="",
        ...rest}=props
    const getUnit=() => {
        return `calc(${gap} * 4pt)`
      }
    return (<GridContainerWrap unit={getUnit()} wrap={wrap} >
        {children}
    </GridContainerWrap>)
}

export default GridContainer