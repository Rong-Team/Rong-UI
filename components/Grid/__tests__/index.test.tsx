import React from 'react'
import Grid from '../index'
import { mount } from 'enzyme'
import { shallowWithTheme } from '../../utils/testUtils'

describe('Grid Test', () => {
    it("render grid",()=>{
        const wrap=shallowWithTheme( <Grid.Container gap={2} justify="center">
        <Grid xs={6}><div style={{width:"60px",height:"60px",background:"red"}}></div></Grid>
        <Grid xs={6}><div style={{width:"60px",height:"60px",background:"red"}}></div></Grid>
        <Grid xs={6}><div style={{width:"60px",height:"60px",background:"red"}}></div></Grid>
      </Grid.Container>)
      expect(wrap).toMatchSnapshot()
    })
})
