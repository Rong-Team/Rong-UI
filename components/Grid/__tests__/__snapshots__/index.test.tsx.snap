// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid Test render grid 1`] = `ShallowWrapper {}`;
